### PR TITLE
Simplify bank_forks, add squashing and pruning

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
1) Removing parents of new banks from bank_forks on insert() makes using bank_forks clumsy whenever we want to perform operations/searches on a set of banks (always have to search parents of the stored set).

2) Banks are not squashed after max_lockout is reached in locktower

#### Summary of Changes
1) Stop pruning parents in bank_forks() on insert
2) Add squashing and pruning logic to bank_forks after we reach max lockout on a slot.

Thanks @rob-solana for the discussion.

Fixes #

### CODE REVIEW

File: core/src/bank_forks.rs [hunk 1]
```diff
@@ -104,28 +86,37 @@ impl BankForks {
 
     // TODO: use the bank's own ID instead of receiving a parameter?
```
Comment (rob-solana) [inline core/src/bank_forks.rs]: can you clean up this TODO while you're here?

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -859,6 +859,24 @@ impl Bank {
         self.accounts
             .store_slow(self.accounts_id, &program_id, &bogus_account);
     }
+
+    pub fn is_descendant_of(&self, parent: u64) -> bool {
```
Comment (rob-solana) [inline runtime/src/bank.rs]: can you think of a better name since bank0.is_descendant_of(0) == true?

File: runtime/src/bank.rs [hunk 3]
```diff
@@ -1624,4 +1642,28 @@ mod tests {
...
+        // Bank 5
+        let bank5 = Bank::new_from_parent(&bank, &Pubkey::default(), 5);
+
+        // Parents of bank 2: 0 -> 1 -> 2
+        assert!(bank2.is_descendant_of(0));
+        assert!(bank2.is_descendant_of(1));
+        assert!(bank2.is_descendant_of(2));
+        assert!(!bank2.is_descendant_of(3));
+
+        // Parents of bank 3: 0 -> 1 -> 5
```
Comment (rob-solana) [inline runtime/src/bank.rs]: comment is wrong?

File: core/src/bank_forks.rs [hunk 4]
```diff
@@ -104,28 +86,37 @@ impl BankForks {
...
+            .banks
+            .get(&root)
+            .expect("root bank didn't exist in bank_forks");
+        root_bank.squash();
+        self.prune_non_root(root);
+    }
+
+    fn prune_non_root(&mut self, root: u64) {
+        self.banks.retain(|slot, bank| {
+            if *slot < root {
```
Comment (garious) [inline core/src/bank_forks.rs]: How about:
```rust
self.banks.retain(|slot, bank| *slot >= root || bank.is_descendant_of(root))
```

File: core/src/replay_stage.rs [hunk 5]
```diff
@@ -576,4 +576,40 @@ mod test {
...
+            // Insert blob for slot 1, generate new forks, check result
+            let mut blob_slot_1 = Blob::default();
+            blob_slot_1.set_slot(1);
+            blob_slot_1.set_parent(0);
+            blocktree.insert_data_blobs(&vec![blob_slot_1]).unwrap();
+            assert!(bank_forks.get(1).is_none());
+            ReplayStage::generate_new_bank_forks(&blocktree, &mut bank_forks);
+            assert!(bank_forks.get(1).is_some());
+
+            // Inset blob for slot 3, generate new forks, check result
```
Comment (garious) [inline core/src/replay_stage.rs]: Insert

File: core/src/bank_forks.rs (hunk 4 above)
Comment (carllin) [reply core/src/bank_forks.rs]: doh! so much better xD

**Pull request merged**